keras_spiking/regularizers.py

```diff
@@ -65,137 +139,127 @@ class L2(L1L2):
-                   spiking_aware_training=False,
-                   activity_regularizer=keras_spiking.PercentileRange(
-                       scale=0.01, minimum=5, maximum=10, penalty="l2"
-                   ),
-               ),
-               tf.keras.layers.GlobalAveragePooling1D(),
-               tf.keras.layers.Dense(1),
-           ]
-       )
-       model.compile(
-           optimizer="rmsprop",
-           loss="mse",
-       )
-       model.fit(train_seq, train_y, epochs=5)
-
-       # create functional model to return spiking layer outputs too
-       func_model = tf.keras.Model(model.input, [model.output, model.layers[1].output])
-       outputs, spikes = func_model.predict(test_seq)
-
-       rates = spikes.mean(axis=1)  # estimate rates by averaging over time
-       max_rates = rates.max(axis=0)
-       print("Max rates: %s, %s" % (max_rates.mean(), max_rates.std()))
-
-       error = np.mean(np.abs(outputs - test_y))
-       print("Error: %s" % (error,))
+        train_x = np.random.uniform(-1, 1, size=(1024 * 100, 2))
+        train_y = train_x[:, :1] * train_x[:, 1:]
+        test_x = np.random.uniform(-1, 1, size=(128, 2))
+        test_y = test_x[:, :1] * test_x[:, 1:]
+
+        # train using one timestep, to speed things up
+        train_seq = train_x[:, None]
+
+        # test using 10 timesteps
+        n_steps = 10
+        test_seq = np.tile(test_x[:, None], (1, n_steps, 1))
+
+        inp = x = tf.keras.Input((None, 2))
+        x = tf.keras.layers.TimeDistributed(tf.keras.layers.Dense(50))(x)
+        x = spikes = keras_spiking.SpikingActivation(
+            "relu",
+            dt=1,
+            activity_regularizer=keras_spiking.regularizers.Percentile(target=(5, 10)),
+        )(x)
+        x = tf.keras.layers.GlobalAveragePooling1D()(x)
+        x = tf.keras.layers.Dense(1)(x)
+
+        model = tf.keras.Model(inp, (x, spikes))
+
+        model.compile(
+            optimizer="rmsprop", loss={"dense_1": "mse"}, metrics={"dense_1": "mae"}
+        )
+        model.fit(train_seq, train_y, epochs=5)
+
+        outputs, spikes = model.predict(test_seq)
+
+        # estimate rates by averaging over time
+        rates = spikes.mean(axis=1)
+        max_rates = rates.max(axis=0)
+        print("Max rates: %s, %s" % (max_rates.mean(), max_rates.std()))
+
+        error = np.mean(np.abs(outputs - test_y))
+        print("MAE: %s" % (error,))
+
 
     .. testoutput::
        :hide:
-       :options: +ELLIPSIS, +NORMALIZE_WHITESPACE
 
        ...
        Max rates: ...
-       Error: ...
-
-    We use ``spiking_aware_training=False``, since activity regularizers perform better
-    on continuous activations (firing rates) than they do on discrete spikes.
+       MAE: ...
     """
 
     def __init__(
-        self, scale=0.01, minimum=0, maximum=0, percentile=100, axis=0, penalty="l2"
+        self,
+        percentile=100,
+        axis=0,
+        target=0,
+        regularizer=tf.keras.regularizers.L2(0.01),
```

Comment:
This does make it more annoying to change the weight on the regularization, since the user has to pass a full regularizer class. I'm wondering if we should just force this to be `L1L2` and then make `l1` and `l2` as parameters to `Percentile`.

Reply:
Yeah that sounds good. I think it's pretty uncommon for people to use something other than L1/L2, so the added convenience is probably worth the loss of that small potential functionality.